src/xrpld/app/tx/detail/AMMClawback.cpp

```diff
@@ -116,12 +116,43 @@ AMMClawback::preclaim(PreclaimContext const& ctx)
     }
 
     std::uint32_t const issuerFlagsIn = sleIssuer->getFieldU32(sfFlags);
-
-    // If AllowTrustLineClawback is not set or NoFreeze is set, return no
-    // permission
-    if (!(issuerFlagsIn & lsfAllowTrustLineClawback) ||
-        (issuerFlagsIn & lsfNoFreeze))
+    if (!(issuerFlagsIn & lsfAllowTrustLineClawback))
+    {
+        JLOG(ctx.j.debug())
+            << "AMM Clawback: Issuer does not set asfAllowTrustLineClawback.";
         return tecNO_PERMISSION;
+    }
+
+    if (asset.holds<Issue>())
+    {
+        if (issuerFlagsIn & lsfNoFreeze)
+        {
+            JLOG(ctx.j.debug()) << "AMM Clawback: Issuer sets lsfNoFreeze, "
+                                   "clawback is not allowed.";
+            return tecNO_PERMISSION;
+        }
+    }
+    else if (asset.holds<MPTIssue>())
+    {
+        auto const issuanceKey =
+            keylet::mptIssuance(asset.get<MPTIssue>().getMptID());
+        auto const sleIssuance = ctx.view.read(issuanceKey);
+        if (!sleIssuance)
+        {
+            // this should not happen, because the MPTIssue should exist if
+            // there's any holder
+            return tecOBJECT_NOT_FOUND;  // LCOV_EXCL_LINE
+        }
+
+        if (!((*sleIssuance)[sfFlags] & lsfMPTCanClawback))
```

Comment:
Should we also check if `lsfMPTCanLock` is set similar to it's done for `Issue`?